kernel/safe_typing.ml

```diff
@@ -892,6 +883,60 @@ let add_constant l decl senv =
 let add_constant ?typing_flags l decl senv =
   with_typing_flags ?typing_flags senv ~f:(add_constant l decl)
 
+type opaque_certificate = {
+  opq_body : Constr.t;
+  opq_univs : Univ.ContextSet.t Opaqueproof.delayed_universes;
+  opq_handle : Opaqueproof.opaque_handle;
+  opq_nonce : Nonce.t;
+}
+
+let check_opaque senv (i : Opaqueproof.opaque_handle) pf =
+  let ty_ctx, trust, nonce =
+    try HandleMap.find i senv.future_cst
+    with Not_found ->
+      CErrors.anomaly Pp.(str "Missing opaque with identifier " ++ int (Opaqueproof.repr_handle i))
+  in
+  let handle env body eff =
+    let body, uctx, signatures, skip = inline_side_effects env body eff in
+    let trusted = check_signatures trust signatures in
+    let trusted, uctx = match trusted with
+    | None -> 0, uctx
+    | Some univs -> skip, Univ.ContextSet.union univs uctx
+    in
+    body, uctx, trusted
+  in
+  let (c, ctx) = Term_typing.check_delayed handle ty_ctx pf in
+  let c = Constr.hcons c in
+  let ctx = match ctx with
+  | Opaqueproof.PrivateMonomorphic u ->
+    Opaqueproof.PrivateMonomorphic (Univ.hcons_universe_context_set u)
+  | Opaqueproof.PrivatePolymorphic (n, u) ->
+    Opaqueproof.PrivatePolymorphic (n, Univ.hcons_universe_context_set u)
+  in
+  { opq_body = c; opq_univs = ctx; opq_handle = i; opq_nonce = nonce }
+
+let fill_opaque { opq_univs = ctx; opq_handle = i; opq_nonce = n; _ } senv =
+match HandleMap.find i senv.future_cst with
+| _, _, nonce ->
+  let () =
+    if not (Nonce.equal n nonce) then
+      CErrors.anomaly  Pp.(str "Invalid opaque certificate")
+  in
+  (* TODO: Drop the the monomorphic constraints, they should really be internal
+     but the higher levels use them haphazardly. *)
+  let senv = match ctx with
+  | Opaqueproof.PrivateMonomorphic ctx -> add_constraints ctx senv
+  | Opaqueproof.PrivatePolymorphic _ -> senv
+  in
+  (* Mark the constant as having been checked *)
+  { senv with future_cst = HandleMap.remove i senv.future_cst }
+| exception Not_found ->
+  (* TODO: give a better API *)
+  senv
```

Comment:
When do we get Not_found here? Should we anomaly instead of returning the input env?

Reply:
No, this code path is exercised. The problem is that the join operation is not in the CVS tree, so if you do it twice you'll get an anomaly. Joining once is essentially introducing a desynchronization between the safe env and the opaque table, but it does not matter because the only discrepancy is that the opaque table forgets that the safe env has been filled. To fix this we should mark somehow that a join was performed somewhere in the structure so that it is remember correctly.

Reply:
The comment is precisely about that, since I don't know of the best API to perform the double check. The various ways I see are: returning an option / providing a `is_filled` primitive / raising a particular exception. All of these choices have some drawbacks and I can't decide for any. Also there is only one caller of `fill_opaque` so it's somehow introducing quite a tight coupling. I'll try to write a small enhancement anyways.